File: attocode/attocode_py/src/attocode/shared/shared_economics_state.py
"""Shared economics state for cross-worker doom loop detection.

Tracks tool fingerprints globally across all swarm workers to detect
doom loops that span multiple agents.
"""

from __future__ import annotations

import threading
from collections import defaultdict
from dataclasses import dataclass, field
from typing import Any


@dataclass(slots=True)
class SharedEconomicsConfig:
    """Configuration for shared economics state."""

    global_doom_threshold: int = 10


@dataclass(slots=True)
class GlobalLoopInfo:
    """Information about a global tool call pattern."""

    fingerprint: str
    total_calls: int
    worker_count: int
    workers: list[str]


@dataclass
class SharedEconomicsState:
    """Cross-worker doom loop detection and aggregation.

    Tracks tool fingerprints from all workers to detect global patterns
    like multiple workers hitting the same error or calling the same
    tool with identical arguments.
    """

    config: SharedEconomicsConfig = field(default_factory=SharedEconomicsConfig)

    # fingerprint -> {worker_id -> count}
    _global_calls: dict[str, dict[str, int]] = field(
        default_factory=lambda: defaultdict(lambda: defaultdict(int))
    )
    _lock: threading.Lock = field(default_factory=threading.Lock)

    def record_tool_call(self, worker_id: str, fingerprint: str) -> None:
        """Report a tool call to the shared pool."""
        with self._lock:
            self._global_calls[fingerprint][worker_id] += 1

    def is_global_doom_loop(self, fingerprint: str) -> bool:
        """Check if fingerprint exceeded global threshold."""
        with self._lock:
            if fingerprint not in self._global_calls:
                return False
            total = sum(self._global_calls[fingerprint].values())
            return total >= self.config.global_doom_threshold

    def get_global_loop_info(self, fingerprint: str) -> GlobalLoopInfo | None:
        """Get call count and worker count for a fingerprint."""
        with self._lock:
            if fingerprint not in self._global_calls:
                return None
            workers = self._global_calls[fingerprint]
            return GlobalLoopInfo(
                fingerprint=fingerprint,
                total_calls=sum(workers.values()),
                worker_count=len(workers),
                workers=list(workers.keys()),
            )

    def get_global_loops(self) -> list[GlobalLoopInfo]:
        """List all fingerprints exceeding threshold."""
        with self._lock:
            results = []
            for fp, workers in self._global_calls.items():
                total = sum(workers.values())
                if total >= self.config.global_doom_threshold:
                    results.append(GlobalLoopInfo(
                        fingerprint=fp,
                        total_calls=total,
                        worker_count=len(workers),
                        workers=list(workers.keys()),
                    ))
            return results

    def get_stats(self) -> dict[str, int]:
        """Get aggregated stats."""
        with self._lock:
            unique_fingerprints = len(self._global_calls)
            total_calls = sum(
                sum(w.values()) for w in self._global_calls.values()
            )
            loops = sum(
                1 for w in self._global_calls.values()
                if sum(w.values()) >= self.config.global_doom_threshold
            )
            return {
                "unique_fingerprints": unique_fingerprints,
                "total_calls": total_calls,
                "active_doom_loops": loops,
            }

    def to_json(self) -> dict[str, Any]:
        """Serialize for checkpoint persistence."""
        with self._lock:
            return {
                "global_calls": {
                    fp: dict(workers)
                    for fp, workers in self._global_calls.items()
                },
            }

    def restore_from(self, data: dict[str, Any]) -> None:
        """Restore from checkpoint data."""
        with self._lock:
            self._global_calls = defaultdict(lambda: defaultdict(int))
            for fp, workers in data.get("global_calls", {}).items():
                for w, count in workers.items():
                    self._global_calls[fp][w] = count

    def clear(self) -> None:
        """Clear all state."""
        with self._lock:
            self._global_calls.clear()

File: legacy/src/shared/shared-economics-state.ts
/**
 * Shared Economics State
 *
 * Cross-worker doom loop aggregation for swarm execution.
 * Detects when multiple workers are stuck on the same tool call pattern,
 * even though each worker's local doom loop detector only sees its own calls.
 *
 * Reuses computeToolFingerprint from economics.ts.
 */

// =============================================================================
// TYPES
// =============================================================================

export interface SharedEconomicsConfig {
  /** Threshold for global doom loop detection across all workers (default: 10) */
  globalDoomLoopThreshold?: number;
}

interface FingerprintEntry {
  count: number;
  workers: Set<string>;
}

// =============================================================================
// SHARED ECONOMICS STATE
// =============================================================================

export class SharedEconomicsState {
  private toolFingerprints: Map<string, FingerprintEntry> = new Map();
  private threshold: number;

  constructor(config?: SharedEconomicsConfig) {
    this.threshold = config?.globalDoomLoopThreshold ?? 10;
  }

  /**
   * Record a tool call fingerprint from a worker.
   * Called by per-worker ExecutionEconomicsManager.recordToolCall().
   */
  recordToolCall(workerId: string, fingerprint: string): void {
    let entry = this.toolFingerprints.get(fingerprint);
    if (!entry) {
      entry = { count: 0, workers: new Set() };
      this.toolFingerprints.set(fingerprint, entry);
    }
    entry.count++;
    entry.workers.add(workerId);
  }

  /**
   * Check if a tool fingerprint has exceeded the global doom loop threshold.
   */
  isGlobalDoomLoop(fingerprint: string): boolean {
    const entry = this.toolFingerprints.get(fingerprint);
    if (!entry) return false;
    return entry.count >= this.threshold;
  }

  /**
   * Get detailed info about a fingerprint's global usage.
   */
  getGlobalLoopInfo(fingerprint: string): { count: number; workerCount: number } | null {
    const entry = this.toolFingerprints.get(fingerprint);
    if (!entry) return null;
    return { count: entry.count, workerCount: entry.workers.size };
  }

  /**
   * Get all fingerprints that have exceeded the global threshold.
   */
  getGlobalLoops(): string[] {
    const loops: string[] = [];
    for (const [fingerprint, entry] of this.toolFingerprints) {
      if (entry.count >= this.threshold) {
        loops.push(fingerprint);
      }
    }
    return loops;
  }

  // ---------------------------------------------------------------------------
  // Serialization (for checkpoint persistence)
  // ---------------------------------------------------------------------------

  /** Serialize state for checkpoint persistence. */
  toJSON(): { fingerprints: Array<{ fingerprint: string; count: number; workers: string[] }> } {
    const entries: Array<{ fingerprint: string; count: number; workers: string[] }> = [];
    for (const [fp, entry] of this.toolFingerprints) {
      entries.push({ fingerprint: fp, count: entry.count, workers: Array.from(entry.workers) });
    }
    return { fingerprints: entries };
  }

  /** Restore state from checkpoint data. */
  restoreFrom(data: {
    fingerprints?: Array<{ fingerprint: string; count: number; workers: string[] }>;
  }): void {
    if (data.fingerprints) {
      for (const entry of data.fingerprints) {
        this.toolFingerprints.set(entry.fingerprint, {
          count: entry.count,
          workers: new Set(entry.workers),
        });
      }
    }
  }

  getStats(): { fingerprints: number; globalLoops: string[] } {
    return {
      fingerprints: this.toolFingerprints.size,
      globalLoops: this.getGlobalLoops(),
    };
  }

  clear(): void {
    this.toolFingerprints.clear();
  }
}

// =============================================================================
// FACTORY
// =============================================================================

export function createSharedEconomicsState(config?: SharedEconomicsConfig): SharedEconomicsState {
  return new SharedEconomicsState(config);
}
